File: spicedb-typescript/src/__tests__/deadline.test.ts
import { describe, it, expect } from "vitest";
import { createRouterTransport } from "@connectrpc/connect";
import { create } from "@bufbuild/protobuf";
import {
  SpiceDBProtoClient,
  PermissionsService,
  CheckBulkPermissionsResponseSchema,
  ReadRelationshipsResponseSchema,
  RelationshipSchema,
  ObjectReferenceSchema,
  SubjectReferenceSchema,
} from "@spicedb/proto";
import { SpiceDBClient, type SpiceDBClientOptions } from "../client.js";
import { full } from "../consistency.js";
import { DeadlineExceededError } from "../errors.js";

// ---------------------------------------------------------------------------
// Call-deadline enforcement. Root DESIGN.md, "RULE: A unary call must have a
// deadline".
//
// Uses `createRouterTransport` (from `@connectrpc/connect`) instead of a
// mocked `proto` field (the convention elsewhere in this suite): deadline
// enforcement -- `CallOptions.timeoutMs` racing an `AbortController` against
// the call, in `protocol/signals.js`'s `createDeadlineSignal` -- lives in
// Connect's own transport machinery, which a `vi.fn()` mock bypasses
// entirely. `createRouterTransport` runs the real client-side transport
// stack (the same one `createGrpcTransport` uses in production) in-process,
// against a real handler, so these tests exercise genuine enforcement
// without a real socket.
//
// Every stalling handler self-bounds its stall to STALL_MS rather than
// blocking forever: long enough to dwarf the tiny per-test deadlines below
// (proving enforcement, not luck). Each call is additionally wrapped in
// `withWatchdog`, which fails the test (instead of hanging it, and CI along
// with it) if a regression reintroduces an unbounded call.
// ---------------------------------------------------------------------------

const STALL_MS = 2000;
const WATCHDOG_MS = 10_000;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function withWatchdog<T>(p: Promise<T>, ms = WATCHDOG_MS): Promise<T> {
  return Promise.race([
    p,
    new Promise<T>((_, reject) =>
      setTimeout(
        () =>
          reject(
            new Error(
              `call did not settle within ${ms}ms -- deadline enforcement regressed`,
            ),
          ),
        ms,
      ),
    ),
  ]);
}

function clientWithTransport(
  transport: ReturnType<typeof createRouterTransport>,
  options: Partial<SpiceDBClientOptions> = {},
): SpiceDBClient {
  const client = new SpiceDBClient({
    endpoint: "localhost:1",
    token: "test-token",
    insecure: true,
    ...options,
  });
  (client as unknown as { proto: SpiceDBProtoClient }).proto =
    new SpiceDBProtoClient(transport);
  return client;
}

const check = {
  resourceType: "document",
  resourceId: "1",
  permission: "view",
  subjectType: "user",
  subjectId: "alice",
};

describe("call deadlines", () => {
  it("fails a unary call against a stub that never responds with DeadlineExceededError, not a hang", async () => {
    const transport = createRouterTransport((router) => {
      router.service(PermissionsService, {
        checkBulkPermissions: async () => {
          await sleep(STALL_MS);
          return create(CheckBulkPermissionsResponseSchema, {});
        },
      });
    });
    const client = clientWithTransport(transport, { defaultTimeoutMs: 200 });

    const start = Date.now();
    await expect(
      withWatchdog(client.checkPermissions(full(), [check])),
    ).rejects.toBeInstanceOf(DeadlineExceededError);
    const elapsed = Date.now() - start;

    expect(elapsed).toBeLessThan(STALL_MS);
  });

  it("lets a per-call timeoutMs override a much larger client default", async () => {
    const transport = createRouterTransport((router) => {
      router.service(PermissionsService, {
        checkBulkPermissions: async () => {
          await sleep(STALL_MS);
          return create(CheckBulkPermissionsResponseSchema, {});
        },
      });
    });
    // Client default is larger than the server's stall -- if the per-call
    // override did not take effect, this call would not fail quickly.
    const client = clientWithTransport(transport, {
      defaultTimeoutMs: STALL_MS * 10,
    });

    const start = Date.now();
    await expect(
      withWatchdog(
        client.checkPermissions(full(), [check], { timeoutMs: 200 }),
      ),
    ).rejects.toBeInstanceOf(DeadlineExceededError);
    const elapsed = Date.now() - start;

    expect(elapsed).toBeLessThan(STALL_MS);
  });

  it("does not cut a streaming call off at the (tiny) unary default", async () => {
    const transport = createRouterTransport((router) => {
      router.service(PermissionsService, {
        async *readRelationships() {
          await sleep(STALL_MS);
          yield create(ReadRelationshipsResponseSchema, {
            relationship: create(RelationshipSchema, {
              resource: create(ObjectReferenceSchema, {
                objectType: "document",
                objectId: "a",
              }),
              relation: "viewer",
              subject: create(SubjectReferenceSchema, {
                object: create(ObjectReferenceSchema, {
                  objectType: "user",
                  objectId: "jimmy",
                }),
              }),
            }),
          });
        },
      });
    });
    // defaultTimeoutMs is far smaller than the server's stall. If
    // readRelationships inherited it, this would reject with
    // DeadlineExceededError instead of yielding the item.
    const client = clientWithTransport(transport, { defaultTimeoutMs: 100 });

    const start = Date.now();
    const items = await withWatchdog(
      (async () => {
        const collected = [];
        for await (const rel of client.readRelationships(
          { resourceType: "document" },
          full(),
        )) {
          collected.push(rel);
        }
        return collected;
      })(),
    );
    const elapsed = Date.now() - start;

    expect(items.map((r) => r.resourceId)).toEqual(["a"]);
    expect(elapsed).toBeGreaterThanOrEqual(STALL_MS);
  });

  describe("defaultTimeoutMs", () => {
    it("defaults to 30_000, mirroring authzed-node's DEFAULT_DEADLINE_MS", () => {
      const client = new SpiceDBClient({
        endpoint: "localhost:1",
        token: "t",
        insecure: true,
      });
      expect(
        (client as unknown as { defaultTimeoutMs: number }).defaultTimeoutMs,
      ).toBe(30_000);
    });

    it("falls back to the client default when no per-call timeoutMs is given", () => {
      const client = new SpiceDBClient({
        endpoint: "localhost:1",
        token: "t",
        insecure: true,
        defaultTimeoutMs: 7500,
      });
      const effective = (
        client as unknown as {
          effectiveTimeoutMs(timeoutMs?: number): number;
        }
      ).effectiveTimeoutMs.bind(client);
      expect(effective(undefined)).toBe(7500);
      expect(effective(1500)).toBe(1500);
    });
  });
});

File: spicedb-rust/tests/deadline_test.rs
//! Call-deadline enforcement. Root DESIGN.md, "RULE: A unary call must have
//! a deadline".
//!
//! Runs a real in-process gRPC server (`support::MockPermissionsService`)
//! whose handlers deliberately stall, so these tests exercise the real
//! `tonic::Request::set_timeout` plumbing end to end. A mocked/queued
//! response can't prove a deadline is actually enforced -- tonic's deadline
//! machinery lives in the transport layer, below any application-level mock.
//!
//! Every stalling handler self-bounds its stall to `STALL` rather than
//! blocking forever: long enough to dwarf the tiny per-test deadlines below
//! (proving enforcement, not luck), short enough that a test failure doesn't
//! hang the suite for long even without the watchdog. Each call is
//! additionally wrapped in `tokio::time::timeout(WATCHDOG, ...)`, which fails
//! the test (instead of hanging it, and CI along with it) if a regression
//! reintroduces an unbounded call.

mod support;

use std::time::Duration;

use futures::StreamExt;
use spicedb::client::SpiceDBClient;
use spicedb::consistency;
use spicedb::error::SpiceDBError;
use spicedb::types::{Filter, Relationship};
use spicedb_proto::authzed::api::v1 as proto;

use support::{spawn_permissions_server, MockPermissionsService};

/// How long a stalling handler stalls before finally answering. Every
/// per-test deadline below is far smaller than this, so a passing test
/// proves the client's timeout fired -- it did not just get lucky with
/// scheduling.
const STALL: Duration = Duration::from_secs(2);

/// Wall-clock budget for a single watchdog-wrapped call.
const WATCHDOG: Duration = Duration::from_secs(10);

fn has_permission_response() -> proto::CheckBulkPermissionsResponse {
    proto::CheckBulkPermissionsResponse {
        checked_at: Some(proto::ZedToken {
            token: "rev-1".to_string(),
        }),
        pairs: vec![proto::CheckBulkPermissionsPair {
            request: None,
            response: Some(proto::check_bulk_permissions_pair::Response::Item(
                proto::CheckBulkPermissionsResponseItem {
                    permissionship: proto::check_permission_response::Permissionship::HasPermission
                        as i32,
                    partial_caveat_info: None,
                    debug_trace: None,
                },
            )),
        }],
    }
}

#[tokio::test]
async fn unary_call_against_a_stub_that_never_responds_times_out_with_deadline_exceeded() {
    let mock = MockPermissionsService::new();
    mock.stall_check_bulk_permissions(STALL);
    mock.push_check_bulk_permissions_response(has_permission_response());

    let addr = spawn_permissions_server(mock).await;
    let client = SpiceDBClient::builder(addr.to_string(), "token")
        .plaintext()
        .default_timeout(Duration::from_millis(200))
        .build()
        .await
        .expect("client should connect to mock server");

    let rel = Relationship::new("document", "doc1", "view", "user", "alice", "")
        .expect("valid relationship");

    let start = std::time::Instant::now();
    let result = tokio::time::timeout(
        WATCHDOG,
        client.check_permissions(&consistency::full(), "view", &[rel]),
    )
    .await
    .expect("call did not return within the watchdog -- deadline enforcement regressed");
    let elapsed = start.elapsed();

    let err = result.expect_err("a stalling server must not let the call succeed");
    assert!(
        matches!(err, SpiceDBError::DeadlineExceeded(_)),
        "expected SpiceDBError::DeadlineExceeded, got {err:?}"
    );
    assert!(
        elapsed < STALL,
        "the call must fail at the ~200ms client default, not wait out the server's \
         {STALL:?} stall (elapsed={elapsed:?})"
    );
}

#[tokio::test]
async fn per_call_timeout_overrides_a_much_larger_client_default() {
    let mock = MockPermissionsService::new();
    mock.stall_check_bulk_permissions(STALL);
    mock.push_check_bulk_permissions_response(has_permission_response());

    let addr = spawn_permissions_server(mock).await;
    // Client default is larger than the server's stall -- if the per-call
    // override did not take effect, this call would not fail quickly.
    let client = SpiceDBClient::builder(addr.to_string(), "token")
        .plaintext()
        .default_timeout(STALL * 10)
        .build()
        .await
        .expect("client should connect to mock server");

    let rel = Relationship::new("document", "doc1", "view", "user", "alice", "")
        .expect("valid relationship");

    let start = std::time::Instant::now();
    let result = tokio::time::timeout(
        WATCHDOG,
        client.check_permissions_with_timeout(
            &consistency::full(),
            "view",
            &[rel],
            Duration::from_millis(200),
        ),
    )
    .await
    .expect("call did not return within the watchdog -- deadline enforcement regressed");
    let elapsed = start.elapsed();

    let err = result.expect_err("a stalling server must not let the call succeed");
    assert!(
        matches!(err, SpiceDBError::DeadlineExceeded(_)),
        "expected SpiceDBError::DeadlineExceeded, got {err:?}"
    );
    assert!(
        elapsed < STALL,
        "the per-call timeout=200ms must override the large client default (elapsed={elapsed:?})"
    );
}

#[tokio::test]
async fn streaming_call_does_not_inherit_the_unary_default() {
    let mock = MockPermissionsService::new();
    mock.stall_read_relationships(STALL);
    mock.push_read_relationships_page(vec![proto::ReadRelationshipsResponse {
        read_at: None,
        relationship: Some(proto::Relationship {
            resource: Some(proto::ObjectReference {
                object_type: "document".to_string(),
                object_id: "a".to_string(),
            }),
            relation: "viewer".to_string(),
            subject: Some(proto::SubjectReference {
                object: Some(proto::ObjectReference {
                    object_type: "user".to_string(),
                    object_id: "jimmy".to_string(),
                }),
                optional_relation: String::new(),
            }),
            optional_caveat: None,
            optional_expires_at: None,
        }),
        after_result_cursor: None,
    }]);

    let addr = spawn_permissions_server(mock).await;
    // default_timeout is far smaller than the server's stall. If
    // read_relationships inherited it, this would fail with
    // DeadlineExceeded instead of yielding the item.
    let client = SpiceDBClient::builder(addr.to_string(), "token")
        .plaintext()
        .default_timeout(Duration::from_millis(100))
        .build()
        .await
        .expect("client should connect to mock server");

    let filter = Filter::new("document");
    let start = std::time::Instant::now();
    let items: Vec<_> = tokio::time::timeout(WATCHDOG, async {
        let stream = client.read_relationships(&consistency::full(), &filter);
        tokio::pin!(stream);
        stream.collect::<Vec<_>>().await
    })
    .await
    .expect("stream did not complete within the watchdog -- deadline enforcement regressed");
    let elapsed = start.elapsed();

    let resource_ids: Vec<String> = items
        .into_iter()
        .map(|r| r.expect("stream must not fail").resource_id)
        .collect();
    assert_eq!(resource_ids, vec!["a".to_string()]);
    assert!(
        elapsed >= STALL,
        "the stream must outlive the tiny unary default -- it should have waited out the \
         server's {STALL:?} stall (elapsed={elapsed:?})"
    );
}

#[tokio::test]
async fn import_relationships_does_not_inherit_the_unary_default() {
    // import_relationships (ImportBulkRelationships) is client-streaming:
    // its duration scales with the size of the caller's dataset, not with
    // server latency, so root DESIGN.md's "RULE: A unary call must have a
    // deadline" (clause 3, amended to cover client-streaming and
    // bidirectional RPCs) excludes it from `default_timeout`. default_timeout
    // here is far smaller than the server's stall -- if import_relationships
    // inherited it, this call would fail with DeadlineExceeded instead of
    // completing.
    let mock = MockPermissionsService::new();
    mock.stall_import_bulk_relationships(STALL);

    let addr = spawn_permissions_server(mock).await;
    let client = SpiceDBClient::builder(addr.to_string(), "token")
        .plaintext()
        .default_timeout(Duration::from_millis(100))
        .build()
        .await
        .expect("client should connect to mock server");

    let rel = Relationship::new("document", "doc1", "viewer", "user", "alice", "")
        .expect("valid relationship");

    let start = std::time::Instant::now();
    let result = tokio::time::timeout(WATCHDOG, client.import_relationships(vec![rel]))
        .await
        .expect(
            "call did not return within the watchdog -- import_relationships must not inherit \
             the unary default",
        );
    let elapsed = start.elapsed();

    let num_loaded = result.expect("a stalling server must still eventually respond");
    assert_eq!(num_loaded, 1);
    assert!(
        elapsed >= STALL,
        "import_relationships must outlive the tiny unary default -- it should have waited out \
         the server's {STALL:?} stall (elapsed={elapsed:?})"
    );
}

#[tokio::test]
async fn import_relationships_with_timeout_still_bounds_the_call() {
    // The exclusion above is from the *default*, not from the ability to
    // bound the call at all -- import_relationships_with_timeout must still
    // fail quickly against a stalling server.
    let mock = MockPermissionsService::new();
    mock.stall_import_bulk_relationships(STALL);

    let addr = spawn_permissions_server(mock).await;
    let client = SpiceDBClient::builder(addr.to_string(), "token")
        .plaintext()
        .build()
        .await
        .expect("client should connect to mock server");

    let rel = Relationship::new("document", "doc1", "viewer", "user", "alice", "")
        .expect("valid relationship");

    let start = std::time::Instant::now();
    let result = tokio::time::timeout(
        WATCHDOG,
        client.import_relationships_with_timeout(vec![rel], Duration::from_millis(200)),
    )
    .await
    .expect("call did not return within the watchdog -- deadline enforcement regressed");
    let elapsed = start.elapsed();

    let err = result.expect_err("a stalling server must not let the call succeed");
    assert!(
        matches!(err, SpiceDBError::DeadlineExceeded(_)),
        "expected SpiceDBError::DeadlineExceeded, got {err:?}"
    );
    assert!(
        elapsed < STALL,
        "an explicit 200ms timeout on import_relationships_with_timeout must still fire \
         (elapsed={elapsed:?})"
    );
}
